File: item-list/src/main/java/com/gurunars/item_list/FetcherPermutationRange.java
package com.gurunars.item_list;

import java.util.List;

import java8.util.function.BiFunction;


/*
* For two lists returns such a sublist that the remaining parts represent the unchanged portions
* to the left and to the right of this sublist.
*
* E.g. for [1,2,3,4,5,6] and [1,3,4,5,2,6] returns [3,4,5,2]
*
* */
class FetcherPermutationRange<ItemType> implements BiFunction<List<ItemType>, List<ItemType>, FetcherPermutationRange.Range> {

    static class Range {
        private int start, end;

        Range(int start, int end) {
            this.start = start;
            this.end = end;
        }

        int getStart() {
            return start;
        }

        int getEnd() {
            return end;
        }

        @Override
        public String toString() {
            return "R(" + start + ", " + end + ")";
        }
    }

    @Override
    public Range apply(List<ItemType> source, List<ItemType> target) {
        int startOffset, endOffset;

        int sourceSize = source.size();
        int targetSize = target.size();

        int limit = Math.min(sourceSize, targetSize);

        for (startOffset=0; startOffset < limit; startOffset++) {
            if (!source.get(startOffset).equals(target.get(startOffset))) {
                break;
            }
        }

        for (endOffset=0; endOffset< limit; endOffset++) {
            if (!source.get(sourceSize - 1 - endOffset).equals(
                    target.get(targetSize - 1 - endOffset))) {
                break;
            }
        }

        return new Range(startOffset, targetSize-endOffset);
    }

}

File: item-list/src/main/java/com/gurunars/item_list/FetcherPermutationRange.kt
package com.gurunars.item_list


/*
* For two lists returns such a sublist that the remaining parts represent the unchanged portions
* to the left and to the right of this sublist.
*
* E.g. for [1,2,3,4,5,6] and [1,3,4,5,2,6] returns [3,4,5,2]
*
* */
internal class FetcherPermutationRange<ItemType> : BiFunction<List<ItemType>, List<ItemType>, FetcherPermutationRange.Range> {

    internal class Range(val start: Int, val end: Int) {

        override fun toString(): String {
            return "R($start, $end)"
        }
    }

    override fun apply(source: List<ItemType>, target: List<ItemType>): Range {
        var startOffset: Int
        var endOffset: Int

        val sourceSize = source.size
        val targetSize = target.size

        val limit = Math.min(sourceSize, targetSize)

        startOffset = 0
        while (startOffset < limit) {
            if (source[startOffset] != target[startOffset]) {
                break
            }
            startOffset++
        }

        endOffset = 0
        while (endOffset < limit) {
            if (source[sourceSize - 1 - endOffset] != target[targetSize - 1 - endOffset]) {
                break
            }
            endOffset++
        }

        return Range(startOffset, targetSize - endOffset)
    }

}
